Fail attempts to pass anonsub to \$ prototype.

FGasper opened the pull request:
Issue #20384

FGasper committed:
Fail attempts to pass anonsub to \$ prototype.

Issue #20384

demerphq commented on t/comp/proto.t:
```diff
@@ -892,3 +892,16 @@ $_ = sub ($$$$$$$) {};
+    eval 'sub scalarref (\$) { }; scalarref( sub {} )';
+    if ($@ !~ m/anonymous subroutine/ || $@ !~ m/scalarref/) {
+        $err = $@;
+        warn;
```
personally id feel better if this was `print "$_\n" for split /\n/, $@;` or `diag` if its available.

FWIW, if I had written this it would look more like this (untested):
```
my $unexpected_err = "";
my $ok = eval 'sub scalarref (\$) { }; scalarref( sub {} ); 1';
my $err = !$ok ? $@ : "";
if ($err !~ m/anonymous subroutine/ || $err !~ m/scalarref/) {
   $unexpected_err = $err;
}
print( (($ok || $unexpected_err) ? 'not' : '') . "ok $i - anonsub passed to \\\$\n",
   map { "#$_\n" } split /\n/, $unexpected_err );
```
I dont mean to style nit, its probably fine as is, but it makes me leery using $@ in any way other than a plain copy. It also makes me leery not checking that eval returns true, and outputing the error on STDERR when the test output is on STDOUT means that it might be emitted out of order which makes life difficult. Using a single print statement with commented output means it should be output all together in one chunk in the correct order and not upset the harness.  The other changes are just to make intent clear here. When I first read this it was obvious what '$err' "meant" in the code. 

None of this is absolutely necessary to change I suspect. But it would make me feel better if you integrated some of the points I raise here.

FGasper replied:
> It also makes me leery not checking that eval returns true

Does checking `$@` not substitute?

FGasper replied:
Tweaked, @demerphq. Thanks!

demerphq replied:
Historically you couldn't rely on `$@` being true when an eval fails. The only reliable way to tell that you should look at `$@` is if the eval returned a true value.  Maybe this has been fixed enough my concern is irrelevant, but at the same time, using the robust formulation is obviously correct, whereas not using it open up that tiny wedge of doubt. So i just always use the robust formulation, then i don't have to worry. :-)

demerphq commented on t/comp/proto.t:
```diff
@@ -892,3 +892,20 @@ $_ = sub ($$$$$$$) {};
+    else {
+        if ($@ !~ m/anonymous subroutine/ || $@ !~ m/scalarref/) {
+            $failure = $@;
+            print "Unexpected error (or none): $failure";
```
this can confuse the harness. Stick a # at the start, if $failure can be multiline stick a # in front of each.

FGasper replied:
Fixed.

FGasper committed:
Fix & test useless-anonymous-subroutine warnings.

Issue #20384: This fixes a breakage from
35458d36d3c9fcf47064e4379e15276057b0bca5 that Devel::WatchVars’s test suite
caught. A test is added to Perl’s own test suite to prevent regressions.

FGasper commented:
@demerphq I added a 2nd commit to this branch that fixes a related breakage, also reported in the referent issue (#20384). Do you mind looking at it?

Thank you!

leonerd reviewed (approved):
LGTM
